File: src/client/formula/src/evaluate.ts
import type { Expr } from "./parser";
import { type FormulaValue, isFormulaError } from "./types";
import { finite, validateResolverOutput } from "./internal";

/** Structural recursion over `Expr`. Resolver errors and non-finite arithmetic
 * results both short-circuit as `FormulaValue` errors — the evaluator never
 * throws and never returns `NaN`/`Infinity`. Operands evaluate
 * left-to-right; the FIRST error encountered wins. Recursion depth is bounded
 * indirectly by `MAX_AST_NODES`, not `MAX_PARSE_DEPTH`: a flat additive/
 * multiplicative chain (e.g. `a + b + c + ...`) builds one `evaluate` stack
 * frame per node while the parser's own depth counter — which only advances
 * at structural nesting boundaries (parens, calls, unary-minus) — never
 * increments across that chain, so `MAX_PARSE_DEPTH` does not bound it.
 * @param expr An AST produced by `parseFormula`.
 * @param resolve Consumer callback resolving a dotted ref path (e.g.
 * `["hp","max"]`) to a `FormulaValue`. May throw; a thrown value is caught
 * and converted to a `"resolver-error"` rather than propagating.
 * @returns The evaluated number, or the first `FormulaError` encountered.
 * @example
 * ```ts
 * import { evaluate, parseFormula } from "@shadowcat/formula";
 *
 * const expr = parseFormula("1 + hp.max");
 * if (!("error" in expr)) {
 *   evaluate(expr, (path) => (path.join(".") === "hp.max" ? 10 : { error: "unknown-ref", detail: path.join(".") }));
 * }
 * ```
 */
export function evaluate(
  expr: Expr,
  resolve: (path: string[]) => FormulaValue,
): FormulaValue {
  switch (expr.kind) {
    case "num":
      return expr.value;

    case "ref": {
      let v: unknown;
      try {
        v = resolve(expr.path);
      } catch {
        // Never interpolate the caught exception's message: `FormulaError.detail` is
        // player-presentable, and a consumer resolver's thrown
        // message is an internal implementation detail, not for players.
        return {
          error: "resolver-error",
          detail: `resolver threw for '${expr.path.join(".")}'`,
        };
      }
      return validateResolverOutput(v);
    }

    case "neg": {
      const v = evaluate(expr.operand, resolve);
      if (isFormulaError(v)) return v;
      return finite(-v);
    }

    case "bin": {
      const left = evaluate(expr.left, resolve);
      if (isFormulaError(left)) return left;
      const right = evaluate(expr.right, resolve);
      if (isFormulaError(right)) return right;
      return evalBin(expr.op, left, right);
    }

    case "call":
      return evalCall(expr.fn, expr.args, resolve);
  }
}

/**
 * Applies one already-evaluated binary operator. `/` is float division and
 * `%` is JS's truncated remainder (`-7 % 2 === -1`, not floored modulo) —
 * neither implicitly rounds. `x / 0` and `x % 0` are both `"div-zero"`
 * regardless of the sign or finiteness of `x`; every other result passes
 * through `finite()`, so an overflow (e.g. two huge operands multiplied)
 * becomes `"non-finite"` rather than `Infinity`.
 * @param op One of `+ - * / %`.
 * @param left Left operand, already evaluated to a finite number.
 * @param right Right operand, already evaluated to a finite number.
 * @returns The arithmetic result, or a `"div-zero"`/`"non-finite"` error.
 * @example
 * ```
 * // not part of the public `@shadowcat/formula` surface (this module is not
 * // re-exported).
 * evalBin("%", -7, 2); // -1 (truncated, not floored)
 * evalBin("/", 1, 0);  // { error: "div-zero", detail: "..." }
 * ```
 */
function evalBin(
  op: "+" | "-" | "*" | "/" | "%",
  left: number,
  right: number,
): FormulaValue {
  if ((op === "/" || op === "%") && right === 0) {
    return { error: "div-zero", detail: `division by zero (${op === "/" ? "'/'" : "'%'"})` };
  }
  let result: number;
  switch (op) {
    case "+":
      result = left + right;
      break;
    case "-":
      result = left - right;
      break;
    case "*":
      result = left * right;
      break;
    case "/":
      // Float division per Global Constraints — no implicit rounding.
      result = left / right;
      break;
    case "%":
      // JS truncated remainder (not floored modulo).
      result = left % right;
      break;
  }
  return finite(result);
}

/** Arity is a caller obligation, not one enforced here: today only
 * `parseFormula`'s `checkArity` validates it, at parse time. `Expr` and
 * `evaluate` are public API, so a hand-constructed `Expr` (bypassing the
 * parser) with the wrong argument count is not defended against in this
 * function — `vals[0]` would read `undefined` for a no-arg `floor`/`ceil`/
 * `round`, which `finite()` then rejects as non-finite rather than crashing.
 * Likewise a hand-constructed zero-arg `min`/`max` call spreads an empty
 * array into `Math.min()`/`Math.max()`, which JS defines as `Infinity`/
 * `-Infinity` respectively — also caught by `finite()`, not a crash.
 * @param fn One of `min max floor ceil round`.
 * @param args Argument expressions, evaluated left-to-right before the call
 * (so an error in an earlier argument is returned before a later one is evaluated).
 * @param resolve Consumer callback forwarded to each argument's `evaluate`.
 * @returns The call's numeric result, or the first error among its
 * arguments, or a `"non-finite"` error (see the arity note above).
 * @example
 * ```
 * // not part of the public `@shadowcat/formula` surface (this module is not
 * // re-exported) — reachable only through `evaluate`'s "call" case.
 * evalCall("round", [{ kind: "num", value: -2.5 }], () => 0); // -2 (ties toward +Infinity)
 * ```
 */
function evalCall(
  fn: "min" | "max" | "floor" | "ceil" | "round",
  args: Expr[],
  resolve: (path: string[]) => FormulaValue,
): FormulaValue {
  const vals: number[] = [];
  for (const arg of args) {
    const v = evaluate(arg, resolve);
    if (isFormulaError(v)) return v;
    vals.push(v);
  }
  let result: number;
  switch (fn) {
    case "min":
      result = Math.min(...vals);
      break;
    case "max":
      result = Math.max(...vals);
      break;
    case "floor":
      result = Math.floor(vals[0]);
      break;
    case "ceil":
      result = Math.ceil(vals[0]);
      break;
    case "round":
      // JS-native tie behavior: Math.round rounds ties toward +Infinity, not
      // away from zero — Math.round(-2.5) === -2 (not -3).
      result = Math.round(vals[0]);
      break;
  }
  return finite(result);
}

File: src/server/src/formula/evaluate.rs
//! Structural recursion over `Expr`. Twin of the client package's
//! `evaluate.ts`: operands evaluate left-to-right and the FIRST error wins;
//! `/` is float division, `%` truncated remainder, `round` ties toward +∞;
//! every arithmetic result passes `finite`, so no infinity or NaN escapes.
//! Recursion depth is bounded by `MAX_AST_NODES` (one frame per node on a
//! left-deep chain), not `MAX_PARSE_DEPTH`.

#![deny(missing_docs)]
#![deny(clippy::missing_docs_in_private_items)]

use super::parser::{BinOp, Expr, FnName};
use super::types::{finite, FormulaError, FormulaErrorKind, FormulaValue};

/// Resolves a dotted reference path to a value. The library assigns the path
/// no meaning; an implementation does.
///
/// # Examples
///
/// ```
/// use shadowcat::formula::evaluate::{evaluate, Resolve};
/// use shadowcat::formula::parser::parse;
/// use shadowcat::formula::types::FormulaValue;
///
/// struct Fixed(f64);
/// impl Resolve for Fixed {
///     fn resolve(&self, _path: &[String]) -> FormulaValue {
///         Ok(self.0)
///     }
/// }
///
/// let ast = parse("hp + 1").unwrap();
/// assert_eq!(evaluate(&ast, &Fixed(9.0)), Ok(10.0));
/// ```
pub trait Resolve {
    /// The value at `path`, or a `FormulaError` (typically `UnknownRef`/`Type`).
    ///
    /// # Examples
    ///
    /// ```
    /// use shadowcat::formula::evaluate::Resolve;
    /// use shadowcat::formula::types::FormulaValue;
    ///
    /// // The blanket `impl<F: Fn(&[String]) -> FormulaValue> Resolve for F`
    /// // lets a plain closure serve as a resolver.
    /// let resolver = |path: &[String]| -> FormulaValue { Ok(path.len() as f64) };
    /// assert_eq!(resolver.resolve(&["a".to_string(), "b".to_string()]), Ok(2.0));
    /// ```
    fn resolve(&self, path: &[String]) -> FormulaValue;
}

impl<F: Fn(&[String]) -> FormulaValue> Resolve for F {
    fn resolve(&self, path: &[String]) -> FormulaValue {
        self(path)
    }
}

/// JavaScript `Math.round`: nearest integer, ties toward +∞, and a negative
/// input that rounds to zero keeps its sign (JS yields `-0`).
pub(crate) fn js_round(x: f64) -> f64 {
    let f = x.floor();
    let r = if x - f >= 0.5 { f + 1.0 } else { f };
    if r == 0.0 && x < 0.0 {
        -0.0
    } else {
        r
    }
}

/// Evaluates `expr` against `resolve`. Never panics.
///
/// # Examples
///
/// ```
/// use shadowcat::formula::evaluate::evaluate;
/// use shadowcat::formula::parser::parse;
/// use shadowcat::formula::types::FormulaValue;
///
/// let ast = parse("2 * (3 + 4)").unwrap();
/// let resolver = |_: &[String]| -> FormulaValue { Ok(0.0) }; // no references in this AST
/// assert_eq!(evaluate(&ast, &resolver), Ok(14.0));
/// ```
pub fn evaluate(expr: &Expr, resolve: &dyn Resolve) -> FormulaValue {
    match expr {
        Expr::Num(v) => Ok(*v),
        // A resolver's own error passes through unchanged; a number it returns
        // is gated exactly like an arithmetic result.
        Expr::Ref(path) => resolve.resolve(path).and_then(finite),
        Expr::Neg(operand) => finite(-evaluate(operand, resolve)?),
        Expr::Bin { op, left, right } => {
            let l = evaluate(left, resolve)?;
            let r = evaluate(right, resolve)?;
            eval_bin(*op, l, r)
        }
        Expr::Call { func, args } => {
            let mut vals = Vec::with_capacity(args.len());
            for a in args {
                vals.push(evaluate(a, resolve)?);
            }
            eval_call(*func, &vals)
        }
    }
}

/// Applies one binary operator to two finite operands.
fn eval_bin(op: BinOp, left: f64, right: f64) -> FormulaValue {
    if matches!(op, BinOp::Div | BinOp::Rem) && right == 0.0 {
        let sym = if op == BinOp::Div { "'/'" } else { "'%'" };
        return Err(FormulaError::new(
            FormulaErrorKind::DivZero,
            format!("division by zero ({sym})"),
        ));
    }
    finite(match op {
        BinOp::Add => left + right,
        BinOp::Sub => left - right,
        BinOp::Mul => left * right,
        BinOp::Div => left / right,
        BinOp::Rem => left % right,
    })
}

/// Applies a builtin to already-evaluated arguments. Arity is the parser's
/// obligation; a hand-built `Expr` with the wrong count reaches `finite`
/// as JavaScript would (`floor()` of a missing argument is NaN, `min()` of
/// nothing is +∞), never a panic.
fn eval_call(func: FnName, vals: &[f64]) -> FormulaValue {
    let first = vals.first().copied().unwrap_or(f64::NAN);
    finite(match func {
        FnName::Min => vals.iter().copied().fold(f64::INFINITY, f64::min),
        FnName::Max => vals.iter().copied().fold(f64::NEG_INFINITY, f64::max),
        FnName::Floor => first.floor(),
        FnName::Ceil => first.ceil(),
        FnName::Round => js_round(first),
    })
}

#[cfg(test)]
mod tests;
